File: 백준/Silver/14501. 퇴사/퇴사.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const N = parseInt(input[0]);
const T = [];
const P = [];
let ans = 0;

for (let i = 1; i <= N; i++) {
    const [t, p] = input[i].split(" ").map(Number);
    T.push(t);
    P.push(p);
}

function dfs(day, total) {
    if (day >= N) {
        ans = Math.max(ans, total);
        return;
    }
    if (day+T[day] <= N) {
        dfs(day+T[day], total+P[day]);
    }
    dfs(day+1, total);
}

dfs(0, 0);
console.log(ans);

File: 백준/Silver/14501. 퇴사/퇴사.py
import sys
input = sys.stdin.readline

N = int(input())
T = []
P = []
ans = 0
for _ in range(N):
    t, p = map(int, input().split())
    T.append(t)
    P.append(p)

def dfs(day, total):
    global ans
    if day >= N:
        ans = max(ans, total)
        return
    if day + T[day] <= N:
        dfs(day + T[day], total + P[day])
        
    dfs(day + 1, total)


dfs(0, 0)
print(ans)
